### ISSUE TITLE
Get app list return error:Connection closed prematurely

### ISSUE THREAD

**Issue by lijie-advantech** [cloudfoundry/cf-java-client#833]
when I call the function :
`cf_operations.applications().list()`

I got the error message like this, version is 3.1.0

```
java.io.IOException: Connection closed prematurely
	at reactor.ipc.netty.http.client.HttpClientOperations.onInboundComplete(HttpClientOperations.java:262)
	at reactor.ipc.netty.channel.ChannelOperations.onHandlerTerminate(ChannelOperations.java:425)
	at reactor.ipc.netty.http.client.HttpClientOperations.onOutboundError(HttpClientOperations.java:503)
	at reactor.ipc.netty.channel.ChannelOperations.onError(ChannelOperations.java:231)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.MonoZip$ZipCoordinator.signalError(MonoZip.java:192)
	at reactor.core.publisher.MonoZip$ZipInner.onError(MonoZip.java:336)
	at reactor.core.publisher.FluxOnAssembly$OnAssemblySubscriber.onError(FluxOnAssembly.java:455)
	at reactor.core.publisher.FluxReplay$SizeBoundReplayBuffer.replayNormal(FluxReplay.java:796)
	at reactor.core.publisher.FluxReplay$SizeBoundReplayBuffer.replay(FluxReplay.java:892)
	at reactor.core.publisher.ReplayProcessor.onError(ReplayProcessor.java:454)
	at reactor.core.publisher.MonoProcessor.drainLoop(MonoProcessor.java:508)
	at reactor.core.publisher.MonoProcessor.onError(MonoProcessor.java:302)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onError(FluxPeek.java:207)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.MonoNext$NextSubscriber.onError(MonoNext.java:87)
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onError(MonoFlatMapMany.java:193)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1332)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onError(FluxPeekFuseable.java:222)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:135)
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:67)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:108)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:115)
	at reactor.core.publisher.FluxUsing$UsingFuseableSubscriber.onNext(FluxUsing.java:345)
	at reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onNext(FluxFilterFuseable.java:104)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onNext(FluxPeekFuseable.java:469)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1092)
	at reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onComplete(MonoReduceSeed.java:150)
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:130)
	at reactor.ipc.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:376)
	at reactor.ipc.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:198)
	at reactor.ipc.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:338)
	at reactor.ipc.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:349)
	at reactor.ipc.netty.http.client.HttpClientOperations.onInboundComplete(HttpClientOperations.java:265)
	at reactor.ipc.netty.channel.ChannelOperations.onHandlerTerminate(ChannelOperations.java:425)
	at reactor.ipc.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:575)
	at reactor.ipc.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:129)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1336)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1127)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1162)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by nebhale** [comment]
The error message states that the server side of the connection, or more likely a network devices between the client and the server has closed the socket before all data had been read from it.

**Comment by nebhale** [comment]
Since there's been no response on this issue in a couple of weeks, I'm going to close it.  If you'd like to see it re-opened, please comment on the issue and I'll reopen it.

**Comment by eaglerainbow** [comment]
I'd like to request opening this issue with providing further details:

[Promregator](https://github.com/promregator/promregator) is an application that uses the cf-java-client. @dschroe reported [an issue at Promregator](https://github.com/promregator/promregator/issues/59), which I believe is very similar to this one here. He is capable of reproducing it quite regularly; I was able to reproduce it having debugging mode enabled once by chance. I am [attaching the log file](https://github.com/cloudfoundry/cf-java-client/files/2041846/promregator-closedprem.log), which I could get hold of. 
The coding which is executed can mainly be reviewed at https://github.com/promregator/promregator/blob/c7f1613d9f332cb2ce1a1c562c18328e03aec2f5/src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java:
* The setup procedure of the connection can be found at https://github.com/promregator/promregator/blob/c7f1613d9f332cb2ce1a1c562c18328e03aec2f5/src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java#L175
* The issue happens already with the first business call, which is triggered by https://github.com/promregator/promregator/blob/c7f1613d9f332cb2ce1a1c562c18328e03aec2f5/src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java#L231. The latter mainly consists of
   ```
   ListOrganizationsRequest orgsRequest = ListOrganizationsRequest.builder().name(orgName).build();
   this.cloudFoundryClient.organizations().list(or);
   ```
* Based on the result returned, further operations would be requested (based the that Mono) afterwards. In sum, we can expect that roughly 12 operations will be triggered.
* The `IOException` is raised on the response of `this.cloudFoundryClient.organizations().list(or)`. You can verify this due to the error message `Retrieval of retrieveOrgId with key ibm@schmoigl-online.de raised a reactor error` which you find in the log, which is triggered by https://github.com/promregator/promregator/blob/c7f1613d9f332cb2ce1a1c562c18328e03aec2f5/src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java#L212

I tried to understand what is happening based on the verbose logs attached. What makes me wondering are the log messages at 2018-05-26 16:44:38.566 and later, which read: egistering close event to pool release: 
```
Acquired [id: 0x691802ab, L:/192.168.1.130:58972 - R:api.eu-gb.bluemix.net/2.23.183.41:443], now 15 active connections
[...]
Acquired [id: 0x3a462b0e, L:/192.168.1.130:58973 - R:api.eu-gb.bluemix.net/2.23.183.41:443], now 16 active connections
[...]
Released [id: 0xdc193d2b, L:/192.168.1.130:58958 ! R:api.eu-gb.bluemix.net/2.23.183.41:443], now 15 active connections
Released [id: 0x2afce4a0, L:/192.168.1.130:58960 ! R:api.eu-gb.bluemix.net/2.23.183.41:443], now 14 active connections
Released [id: 0x2ae976b1, L:/192.168.1.130:58961 ! R:api.eu-gb.bluemix.net/2.23.183.41:443], now 13 active connections
```
This makes me puzzled, because we even did not receive an answer from `this.cloudFoundryClient.organizations().list(or);` yet - even worse, we even did not send the request, yet! The only request which was sent, was the API standard request, which reads `GET / HTTP/1.1`.

Oh, before I forget: Version of `cloudfoundry-client` in use is 3.3.0.RELEASE.

Could you please assist in looking at the logs and help understanding what is going wrong here?

Thanks!
